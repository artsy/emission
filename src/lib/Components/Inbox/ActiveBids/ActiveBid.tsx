import * as React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components/native"

import { TouchableWithoutFeedback } from "react-native"

import { BodyText, MetadataText } from "../Typography"

import colors from "../../../../data/colors"
import SwitchBoard from "../../../NativeModules/SwitchBoard"
import OpaqueImageView from "../../OpaqueImageView"

const Container = styled.View`
  margin: 17px 20px 0;
  height: 80px;
`

const Content = styled.View`
  flex: 1;
  flex-direction: row;
  align-items: center;
`

const ImageView = styled(OpaqueImageView)`
  width: 58px;
  height: 58px;
  border-radius: 4;
`

const MetadataContainer = styled.View`
  margin: 0 15px;
  flex: 1;
  flex-direction: column;
  justify-content: center;
`

const Separator = styled.View`
  height: 1;
  width: 100%;
  background-color: ${colors["gray-regular"]};
  margin-top: 17px;
`

const StatusLabel = styled(MetadataText)`
  color: ${(props: { status: BidStatus }) => {
    switch (props.status) {
      case "winning":
        return colors["green-regular"]
      case "reserve":
        return colors["yellow-regular"]
      case "losing":
        return colors["red-regular"]
    }
  }};
`

type BidStatus = "winning" | "reserve" | "losing"

interface State {
  status: BidStatus
}

class ActiveBid extends React.Component<RelayProps, State> {
  constructor(props) {
    super(props)

    this.state = {
      status: "losing",
    }

    this.handleTap = this.handleTap.bind(this)
  }

  componentDidMount() {
    this.updateStatus()
  }

  updateStatus() {
    const bid = this.props.bid
    const leadingBidder = bid.is_leading_bidder
    const reserveNotMet = bid.active_bid.sale_artwork.reserve_status === "reserve_not_met"

    let status: BidStatus = "losing"
    if (leadingBidder) {
      status = reserveNotMet ? "reserve" : "winning"
    }

    this.setState({ status })
  }

  get statusLabel(): string {
    switch (this.state.status) {
      case "winning":
      case "reserve":
        return "Highest Bid"
      case "losing":
        return "Outbid"
    }
  }

  handleTap() {
    SwitchBoard.presentNavigationViewController(this, this.props.bid.active_bid.sale_artwork.artwork.href)
  }

  render() {
    const bid = this.props.bid.active_bid
    const imageURL = bid.sale_artwork.artwork.image.url
    const lotNumber = bid.sale_artwork.lot_number
    const artistName = bid.sale_artwork.artwork.artist_names

    const headline = `Lot ${lotNumber} Â· ${artistName} `
    const subtitle = `Current Bid: ${bid.max_bid.display} `

    return (
      <TouchableWithoutFeedback onPress={this.handleTap}>
        <Container>
          <Content>
            <ImageView imageURL={imageURL} />
            <MetadataContainer>
              <BodyText>
                {headline}
              </BodyText>
              <BodyText>
                {subtitle}
              </BodyText>
            </MetadataContainer>
            <StatusLabel status={this.state.status}>
              {this.statusLabel}
            </StatusLabel>
          </Content>
          <Separator />
        </Container>
      </TouchableWithoutFeedback>
    )
  }
}

export default createFragmentContainer(
  ActiveBid,
  graphql`
    fragment ActiveBid_bid on LotStanding {
      is_leading_bidder
      active_bid {
        max_bid {
          display
        }
        sale_artwork {
          lot_number
          reserve_status
          artwork {
            href
            image {
              url
            }
            artist_names
          }
        }
      }
    }
  `
)

interface RelayProps {
  bid: {
    is_leading_bidder: boolean | null
    active_bid: {
      max_bid: {
        display: string | null
      } | null
      sale_artwork: {
        lot_number: string | null
        reserve_status: string | null
        artwork: {
          href: string | null
          image: {
            url: string | null
          } | null
          artist_names: string | null
        } | null
      } | null
    } | null
  }
}
