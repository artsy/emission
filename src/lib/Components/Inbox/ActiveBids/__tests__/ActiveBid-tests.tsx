import * as React from "react"
import * as renderer from "react-test-renderer"

import ActiveBid from "../ActiveBid"

it("renders correctly", () => {
  const tree = renderer.create(<ActiveBid bid={bid} />).toJSON()
  expect(tree).toMatchSnapshot()
})

const bid = {
  active_bid: {
    id: "594933e6275b244305851e9c",
    display_max_bid_amount_dollars: "$10,000",
    max_bid: {
      cents: 1000000,
      display: "$10,000",
    },
    sale_artwork: {
      lot_label: "8",
      lot_number: "8",
      position: 8,
      highest_bid: {
        cents: 1000000,
        display: "$10,000",
      },
      artwork: {
        id: "robert-longo-untitled-dividing-time",
        title: "Untitled (Dividing Time)",
        image: {
          image_url: "https://d32dm0rphc51dk.cloudfront.net/4GlhFa7ci5-0W25sjDNFIQ/:version.jpg",
        },
        artist: {
          name: "Robert Longo",
        },
      },
    },
  },
}
