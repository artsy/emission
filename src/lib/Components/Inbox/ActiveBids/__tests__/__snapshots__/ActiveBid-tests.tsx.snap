// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`looks right for bids in live auctions that haven't started yet 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "height": 120,
        "marginLeft": 20,
        "marginRight": 20,
        "width": 708,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    <AROpaqueImageView
      imageURL="https://d32dm0rphc51dk.cloudfront.net/4GlhFa7ci5-0W25sjDNFIQ/large.jpg"
      style={
        Array [
          Object {
            "borderRadius": 4,
            "height": 80,
            "width": 80,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "marginBottom": 0,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 0,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#cccccc",
              "fontSize": 11,
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "#f7625a",
                "marginBottom": 3,
              },
              undefined,
            ],
            Object {
              "fontFamily": "AvantGardeGothicITC",
            },
          ]
        }
      >
        OUTBID
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 16,
              "textAlign": "left",
            },
            undefined,
            Object {},
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }
      >
        Lot 8 · Robert Longo 
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 16,
              "textAlign": "left",
            },
            undefined,
            Object {},
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }
      >
        $10,000 (1 Bid)
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e5e5e5",
          "height": 1,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`looks right for bids in live open auctions 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "height": 120,
        "marginLeft": 20,
        "marginRight": 20,
        "width": 708,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    <AROpaqueImageView
      imageURL="https://d32dm0rphc51dk.cloudfront.net/4GlhFa7ci5-0W25sjDNFIQ/large.jpg"
      style={
        Array [
          Object {
            "borderRadius": 4,
            "height": 80,
            "width": 80,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "marginBottom": 0,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 0,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#cccccc",
              "fontSize": 11,
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "#6e1fff",
                "marginBottom": 3,
              },
              undefined,
            ],
            Object {
              "fontFamily": "AvantGardeGothicITC",
            },
          ]
        }
      >
        JOIN LIVE
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 16,
              "textAlign": "left",
            },
            undefined,
            Object {},
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }
      >
        Lot 8 · Robert Longo 
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 16,
              "textAlign": "left",
            },
            undefined,
            Object {},
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }
      >
        Live bidding now open
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e5e5e5",
          "height": 1,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`renders correctly 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "height": 120,
        "marginLeft": 20,
        "marginRight": 20,
        "width": 708,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    <AROpaqueImageView
      imageURL="https://d32dm0rphc51dk.cloudfront.net/4GlhFa7ci5-0W25sjDNFIQ/large.jpg"
      style={
        Array [
          Object {
            "borderRadius": 4,
            "height": 80,
            "width": 80,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "marginBottom": 0,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 0,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#cccccc",
              "fontSize": 11,
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "#f7625a",
                "marginBottom": 3,
              },
              undefined,
            ],
            Object {
              "fontFamily": "AvantGardeGothicITC",
            },
          ]
        }
      >
        OUTBID
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 16,
              "textAlign": "left",
            },
            undefined,
            Object {},
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }
      >
        Lot 8 · Robert Longo 
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 16,
              "textAlign": "left",
            },
            undefined,
            Object {},
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }
      >
        $10,000 (1 Bid)
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e5e5e5",
          "height": 1,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;
